File: py/239.py
from typing import *
import collections

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        q = collections.deque()

        res = []
        l = r = 0
        
        while r < len(nums):
            while q and nums[q[-1]] < nums[r]:
                q.pop()
            q.append(r)
            while q[0] < l:
                q.popleft()
            
            if r + 1 >= k:
                res.append(nums[q[0]])
                l += 1
            r += 1
        
        return res



solution = Solution()

nums = [1,3,-1,-3,5,3,6,7]
k = 3
print(solution.maxSlidingWindow(nums, k))

nums = [1]
k = 1
print(solution.maxSlidingWindow(nums, k))

File: go/239.go
package main

import "sort"

func maxSlidingWindow(nums []int, k int) []int {
	res := []int{}

	q := []int{}
	l := 0

	for r := range nums {
		for len(q) > 0 && nums[q[len(q)-1]] <= nums[r] {
			q = q[:len(q)-1]
		}
		q = append(q, r)
		for len(q) > 0 && q[0] < l {
			q = q[1:]
		}
		if r < k-1 {
			continue
		}
		if r-l+1 == k {
			res = append(res, nums[q[0]])
		}
		l++
	}

	return res
}
